Задание 1
Импортируйте библиотеки pandas и numpy.
Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn. Создайте датафреймы X и y из этих данных.
Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки
составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42.
Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.
Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.
Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics.


In [1]:
import numpy as np
import pandas as pd

In [2]:
from sklearn.datasets import load_boston
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

In [3]:
boston = load_boston()
boston.keys()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename'])

In [4]:
data = boston['data']
data.shape

(506, 13)

In [5]:
feature_names = boston['feature_names']
feature_names

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

In [6]:
# print(boston['DESCR'])

In [7]:
target = boston['target']
# target
# target[:10]

In [8]:
x = pd.DataFrame(data, columns=feature_names)
x.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [9]:
# x.info()

In [10]:
y = pd.DataFrame(target, columns=['price'])  # мне не понятно откуда берется я'price' где находится этоот ключ?
y.head()

price
0   24.0
1   21.6
2   34.7
3   33.4
4   36.2

In [11]:
# y.info

In [12]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=42)

In [13]:
lr = LinearRegression()

In [14]:
lr.fit(x_train, y_train)

LinearRegression()

In [15]:
y_pred = lr.predict(x_test)
y_pred.shape

(152, 1)

In [16]:
check_test = pd.DataFrame({
    'y_test': y_test['price'],
    'y_pred': y_pred.flatten()
})
check_test.head()

y_test     y_pred
173    23.6  28.648960
274    32.4  36.495014
491    13.6  15.411193
72     22.8  25.403213
452    16.1  18.855280

In [17]:
R2 = r2_score(check_test['y_pred'], check_test['y_test'])
R2

0.6693702691495612

R2 = 0.6693702691495612.  Значение очень далеко от значения 1. Модель отработала не очень хорошо.

In [18]:
# для тренировки пропишу другие метрики из урока
from sklearn.metrics import mean_squared_error, mean_absolute_error
print(f'Среднее квадратичная  ошибка: {mean_squared_error(check_test["y_pred"], check_test["y_test"])}')
print(f'Среднее абсолютная ошибка: {mean_absolute_error(check_test["y_pred"], check_test["y_test"])}')

Среднее квадратичная  ошибка: 21.51744423117737
Среднее абсолютная ошибка: 3.16270987145742


Задание 2
Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble.
Сделайте агрумент n_estimators равным 1000,
max_depth должен быть равен 12 и random_state сделайте равным 42.
Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression,
но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0],
чтобы получить из датафрейма одномерный массив Numpy,
так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.
Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из предыдущего задания.
Напишите в комментариях к коду, какая модель в данном случае работает лучше.


In [19]:
from sklearn.ensemble import RandomForestRegressor

In [20]:
model = RandomForestRegressor(n_estimators=1000, max_depth=12, random_state=42)

In [21]:
model.fit(x_train, y_train.values[:, 0])

RandomForestRegressor(max_depth=12, n_estimators=1000, random_state=42)

In [22]:
y_model_pred = model.predict(x_test)
y_model_pred.shape

(152,)

In [23]:
check_model_test = pd.DataFrame({
    'y_test': y_test['price'],
    'y_pred': y_model_pred
})
check_model_test.head()

y_test     y_pred
173    23.6  22.806412
274    32.4  31.131464
491    13.6  16.339125
72     22.8  23.810726
452    16.1  17.139521

In [24]:
R2 = r2_score(check_model_test['y_pred'], check_test['y_test'])
R2

0.8479049999699443

R2 = 0.8479049999699443. Теперь значение ближе к единице
Вывод: в данной задаче модель класса RandomForestRegressor отработала лучше,
чем модель класса LinerRegression